crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -529,30 +570,25 @@ where
 
         debug!(%last_segment_index, "Identified last segment index");
 
+        // Limit number of pieces to store, its order by Kademlia distance
         let limit = caches
             .backends
             .iter()
             .fold(0usize, |acc, backend| acc + backend.total_capacity as usize);
-        worker_state.heap.clear();
-        // Change limit to number of pieces
-        worker_state.heap.set_limit(limit);
+
+        // Sort pieces by distance such that they are in ascending order and have higher chance of download
+        // overlapping with other processes like node's sync from DSN
+        let mut piece_indices_to_store = BTreeMap::new();
 
         for segment_index in SegmentIndex::ZERO..=last_segment_index {
             for piece_index in segment_index.segment_piece_indexes() {
-                worker_state.heap.insert(KeyWrapper(piece_index));
+                let key = DistanceForKey::new(self.peer_id, piece_index.to_multihash());
+                piece_indices_to_store.insert(key, piece_index);
```

Comment:
Previous code was inserting into a heap, not `BTreeMap`, meaning it was not possible to use outrageous amount of RAM beyond what farmer would store. Now you're just pushing everything into `BTreeMap` without limits, which is both bad in terms of memory usage and in terms of CPU usage, basically the worst from both worlds.

There are two options here:
* Restore previous logic with actual heap that only keeps allowed number of entries
* Change things, but store values in a flat vector and do sorting afterwards

In first case we'll save RAM and potentially use a bunch of CPU, in another case we'll use more RAM, but save a lot of CPU (and probably memory fragmentation as well) by allocating a large vector beforehand and then sorting it once afterwards.

I think a flat `Vec<(DistranceForKey, PieceIndex)>` is better since even for 1T of history that'll be ~39M of neatly allocated entries in a flat vector that we can later quickly sort and turn into `HashMap`. In fact we'll be able to parallelize both construction and sorting of this vector, which is good for performance in case it is taking a substantial amount of time (not the case right now though).

Reply:
Also assuming cache fits fully, there is no need to sort at all

Reply:
I was thinking about getting I should consider tweaking DistanceForKey to include more information like the previous HeapKey (e.g PieceIndex or something else) to provide better error logging and adapt HashMap or even HashSet for better performance (KBucketDistance doesn't impl Hash and doesn't expose internal variable!).

If the DistanceForKey contains a PieceIndex, we can collect the Vec\<DistanceForKey\> directly, and after that sort it and collect it directly into a HashSet without any additional transformations.